fix logic in make_module_dep w.r.t. excluding loads for toolchain & toolchain components

boegel opened the pull request:
This provides a proper fix for the problem that #2135 dances around, and basically undoes @bartoldeman's PR #2091 since it implements a better solution that also does the right thing when `GCC` (i.e. the bundle of `GCCcore` and `binutils`) is used as a toolchain.

Thanks to @rtmclay for helping out with figuring out the actual problem by looking at the Lmod debug output provided by @akesandgren where reloading the `OpenMPI` module failed because the `binutils` module couldn't be found anymore.

The problem was that `load` statements were still being inserted for `GCCcore` and `binutils` when something was being built with the `GCC` toolchain (e.g. `OpenMPI`) when a module naming scheme that enables `expand_toolchain_load` was being used (for example `HierarchicalMNS`).

The existing code that excludes dependencies that extend `$MODULEPATH` failed because `GCC` (i.e. the toolchain being used) was expanded into `GCCcore` + `binutils` before the list of dependencies that extend `$MODULEPATH` are determined via `path_to_top_of_module_tree`, and hence `excluded_deps` was an empty list while it should have included `GCC`.

This is only a problem when the `GCC` toolchain was being used, since both `GCC` and its `GCCcore` component extend `$MODULEPATH`.

By now passing both the toolchain (e.g. `GCC`) and the toolchain components (`GCCcore` + `binutils`) to `path_to_top_of_module_tree`, the `GCC` toolchain itself is correctly excluded as well (and hence so are the toolchain components).

The additional loop that excludes dependencies of dependencies that was added in #2091 by @bartoldeman is no longer needed now.

cc @geimer, @ocaisa, @damianam

akesandgren commented:
This works like a charm :-)

ocaisa commented:
lgtm

bartoldeman commented:
This certainly a more elegant solution! lgtm2

boegel commented:
Thanks for the feedback everyone, going in to be included in EasyBuild v3.1.1.

bartoldeman commented:
I did not check properly... This fix is good for the compiler-only toolchain case where the toolchain used equals the MODULEPATH extender (in fact my hook for ec in expand_toolchain_load(ec=self.cfg) was another way to fix this because our module naming scheme uses that to check for compiler-only toolchains, but your fix is much better than this).

However the original problem fixed in #2091 is still there.
I can reproduce in the same way as in #2091 with HPL-2.1-intel-2016b.eb. It now includes again these two superfluous loads:
```
load("GCCcore/5.4.0")
 
load("binutils/2.26")
```

bartoldeman commented:
How those that work: HPL using intel-2016b expands the dependencies from intel-2016b.eb:
```
dependencies = [
    ('GCCcore', gccver),
    ('binutils', binutilsver, '-GCCcore-%s' % gccver),
    ('icc', compver, gccsuff),
    ('ifort', compver, gccsuff),
    ('impi', '5.1.3.181', '', ('iccifort', '%s%s' % (compver, gccsuff))),
    ('imkl', '11.3.3.210', '', ('iimpi', version)),
]
```
The modules icc, ifort, and impi are eliminated since they expand $MODULEPATH.
However, GCCcore and binutils, which are dependencies of icc and ifort are no longer eliminated.

boegel commented:
@bartoldeman Hmm, ok, I indeed overlooked that... And since I also changed to test case, the tests didn't complain either.

Can you open a new PR to current `develop`, basically re-adding the changes you had in #2091?

Do also include test case from #2091 in the existing `test_make_module_dep_hmns` test, so we retain both test cases (i.e. the one from #2091 and the one I changed it to in this PR).

Do make sure that the original test from #2091 highlights the reintroduced bug; if not, we need a better test case.

boegel commented:
> I am a little worried about reintroducing #2091 because that also means adding #2135 again.

No, it doesn't, #2135 was a workaround for modules that include `load GCCcore` while they shouldn't, while #2091 *avoids* inserting the `load GCCcore` statements in the first place.

So, as long as your modules have been regenerating with the fix included in #2091, you don't need to avoid reloading modules anymore like we do in #2135 (other than maybe to optimise things, but you'll gain very little from it).

boegel commented:
fix was (re)implemented by @bartoldeman in #2152
